### ISSUE TITLE
development proposal: ngx.print with index parameters

### ISSUE THREAD

**Issue by hkmo** [openresty/lua-nginx-module#938]
Hi,
My proposal is related to the following API function:

ngx.print
syntax: ok, err = ngx.print(...)

We are using this function in conent_by_lua phase.
I would like to be able to print starting from an index x to index x+k without having to create a new string.
Today we are doing this by using lua string.sub() and this seems to create a performance hit according to our testing.

I am going to look at the code and see how this could be added.

Please let me know if you have any comment.

Best Regards

**Comment by agentzh** [comment]
@hkmo I suggest we introduce a new API function that simply takes an FFI cdata buffer as argument.

**Comment by hkmo** [comment]
@agentzh what do you mean by FFI cdata buffer? How does this relate to the existing code? How would the ngx.lua interface look like?

**Comment by agentzh** [comment]
@hkmo Something like below, maybe?

```lua
local output = require "ngx.output"  -- this Lua module can sit inside lua-resty-core.
local ffi = require "ffi"  -- this is the standard LuaJIT module

-- below is the cdata buffer which can be shared on the nginx worker level
local buf = ffi.new("char[256]")

local len = math.min(256, #some_lua_string)
ffi.copy(buf, some_lua_string, math.min(256, #some_lua_string))

-- the output.ffi_write() can be implemented atop FFI as well for maximum performance.
local ok, err = output.ffi_write(buf, len)

-- pointer arithmetic also works, for example.
local offset = 128
local ok, err = output.ffi_write(buf + offset, len)
```

**Comment by hkmo** [comment]
@agentzh In your proposal I see a create buffer and a copy ffi.copy() from the lua string to the buffer. So, maybe the overhead is still similar to string.sub()? How is it algorithmically better? Is it because we are copying by chunk of 256 max?

I was thinking about simply adding an extra API function signature print_() or write() in ngx lua module that will accept two extra parameters; offset and length. 

Thanks

**Comment by agentzh** [comment]
@hkmo You can avoid data copying on the Lua land altogether and pass the Lua strings directly with an offset, as in

```
local ffi = require "ffi"
local ffi_cast = ffi.cast
local const_char_ptr = ffi.typeof("const char *")

local ok, err = output.ffi_write(ffi_cast(my_lua_string, const_char_ptr) + offset, len)
```

But we can also add an offset parameter to make the caller code much cleaner, as in

```
local ok, err = output.ffi_write(my_lua_string, offset, len)
```

This is also potentially faster than using the FFI APIs on the caller side.

**Comment by agentzh** [comment]
@hkmo BTW, most of the time, the overhead of `string.sub()` comes from dynamic memory allocations and string hashing/comparison instead of string copying.

**Comment by alubbe** [comment]
I think that api is a great idea. Would it be possible to have an api for
ngx.socket to return something that could be piped into this function,
e.g., to implement a proxy without allocating lua strings at all?

**Comment by agentzh** [comment]
@alubbe Yeah, that was the plan, i.e., adding `ffi_receive` and `ffi_send` functions to the stream-typed cosocket objects as well. Pull requests welcome :)

**Comment by hkmo** [comment]
@agentzh I am trying to sketch below what we could concretely do to avoid the substring when calling ngx.print(my_lua_string) and we want to print from a given offset and length.

The call to "lua" ngx.print(my_lua_string) API corresponds to a call to "C" ngx_http_lua_ngx_print(lua_State *L).
In the current API we have the following code excerpt (from ngx_http_lua_output.c). I do not see a simple way to add an offset and length at this level, since everything is abstracted (in the API) into "lua_State". 

    static int 
    ngx_http_lua_ngx_print(lua_State *L) 
    { 
       dd("calling lua print"); 
       return ngx_http_lua_ngx_echo(L, 0); 
    } 

    static int 
    ngx_http_lua_ngx_echo(lua_State *L, unsigned newline) 
    {     
         nargs = lua_gettop(L); 
         size = 0; 
    
         for (i = 1; i <= nargs; i++) { 
             type = lua_type(L, i); 
             switch (type) { 
             . . .
                case LUA_TSTRING: 
                     lua_tolstring(L, i, &len); 
                     size += len; 
                     break; 
             . . .
	        }
         }
    
         ctx->seen_body_data = 1;  
     
         cl = ngx_http_lua_chain_get_free_buf(r->connection->log, r->pool, 
                                         &ctx->free_bufs, size); 
    
         b = cl->buf;
    
         for (i = 1; i <= nargs; i++) { 
             type = lua_type(L, i); 
             switch (type) { 
                 case LUA_TNUMBER: 
                 case LUA_TSTRING: 
                     p = lua_tolstring(L, i, &len); 
                     b->last = ngx_copy(b->last, (u_char *) p, len); 
                     break; 
                  . . .
 	        }      
         }

         rc = ngx_http_lua_send_chain_link(r, ctx, cl);
    }

Now going back to creating a new module called "output" with the following API function: 
    ffi_write(my_lua_string, offset, len) 
This API will have to somehow call the above ngx_http_lua_ngx_print(lua_State *L) in order to print, right?
So, the question is how to covert from "my_lua_string, offset, len" to "lua_State"?
Or, is there a simpler way without passing via ngx_http_lua_ngx_print(lua_State *L) ?

Thanks

**Comment by agentzh** [comment]
@hkmo No, the `ffi_write` method should never call any of those classic Lua C APIs like `lua_tolstring`, `lua_type`, and `lua_gettop` on the C land since they will never get actually JIT compiled. See the lua-resty-core library for details.

**Comment by agentzh** [comment]
@hkmo The beauty of `ffi_write` is that you can feed it with pure C string buffers, bypassing Lua strings completely. Also, this API function can be fully JIT compiled.

**Comment by hkmo** [comment]
@agentzh I agree, but do we have any lower-level APIs that could be used to implement "output.ffi_wite(lua_string, offset, len)"?

**Comment by agentzh** [comment]
@hkmo The lower-level API is the NGINX C API.

**Comment by agentzh** [comment]
@hkmo I think the plan is to re-implement `ngx.print` and `ngx.say` atop `ffi_write` or anything that `ffi_write` is based on.

**Comment by hkmo** [comment]
What about this?

    local ffi = require "ffi"
    local ffi_cast = ffi.cast
    local const_char_ptr = ffi.typeof("const char *")
    
    function output.ffi_write(lua_string, offset, len)
    
        local buf = ffi_cast(lua_string, const_char_ptr) + offset
        local str = ffi.string(buf, len)
    
        local ok, err = ngx.print(str)
    
        return ok, err
    end

**Comment by agentzh** [comment]
@hkmo Your version is very expensive because

1. `ffi.string` creates a new Lua string.
2. `ngx.print()` is not JIT compiled (yet).

**Comment by hkmo** [comment]
Any other implementation proposal/sketch/hint for this?

**Comment by hkmo** [comment]
I am not very familiar with the code of nginx lua module, but if I understand correctly, we will need to create a new FFI "C" function in lua-nginx-module:
    int 
    ngx_http_lua_ffi_write(const char* str, size_t offset, size_t len);

This new function can probably be added to: 
    
    ngx_http_lua_output.c
    
Then, create the new lua module in lua-resty-core that will use the above FFI function to print a string:
    output.lua:
    
    output.ffi_write(lua_string, offset, len);
    
The implementation of the lua function would be straightforward (FFI cast followed by a call to the above C FFI function). However, the implementation of "ngx_http_lua_ffi_write()" is still to be defined. Is there any existing function that we could inspire from to write this new function?

As a side note, I am not sure if there is already a header file that exposes all the existing FFI functions provided by lua-nginx module. I think this would show clearly the FFI interface.

Regards

**Comment by agentzh** [comment]
@hkmo Just look at all the existing `ngx_http_lua_ffi_*` C functions in the code base for examples. Some implementation details can also be borrowed from the existing ngx.print() implementation.

**Comment by hkmo** [comment]
@agentzh In the existing ngx.print(), I see an allocation of buffer followed by copy into that buffer:
        cl = ngx_http_lua_chain_get_free_buf(..);
        b = cl->buf;
        b->last = ngx_copy(b->last, (u_char *) str, len);
        rc = ngx_http_lua_send_chain_link(r, ctx, cl);
Is there a way to achieve zero copy? or is this something that cannot be avoided?

**Comment by agentzh** [comment]
@hkmo Zero-copy is not possible here since the original Lua string buffer and the NGINX output buffer have very different life times.

**Comment by agentzh** [comment]
@hkmo Memory allocations should be rare there due to the buffer recycling in `ngx_http_lua_chain_get_free_buf`.

**Comment by hkmo** [comment]
@agentzh I have drafted a code change for ffi_write and output.lua. Can you please have a look? should I just past the code here?

**Comment by hkmo** [comment]
created a pull request for ffi_write new API function.
The code change has been tested.
If you have any code review comment, please let me know.
Thanks
